### PR TITLE
More river configuration options

### PR DESCRIPTION
**Pull request by reldred**
This is probably my largest PR yet, and touches a few things:

Control the width of tropic zone generated around rivers: By default this is 5, I added setting to change it up to 12 which creates a tropic zone buffer the same width as around coastlines. This has the added bonus of on large flat maps providing more areas for tropic industries that cannot generate in deserts to appear.

Control whether rivers require generating at the top of hills: Because there is another condition requiring rivers to have higher tiles around their start point this created a catch 22 where rivers very rarely generated. Anecdotally I have noticed that river generation *feels* quicker. I have no conclusive proof for this. Because rivers will also try and start on another river tile as well, it can result in the river generator creating *very* long chained river systems which is quite pleasing to look at and challenging to build around.

Two extra generation settings for rivers beyond 'Many': This can slow down map generation on large maps considerably so have been highlighted in red in the menu. TODO: Find better names for these strings. They are jokingly set to 'A bit more' and 'About a Few'.

Control the minimum length of rivers: This can help with too many short-stubby rivers, and also controls in an offhand kind of way how often Lakes can be seen. Setting this too high for a map that is very rough can result in less rivers generating.

Control the randomness of rivers: This had zero felt effect on generation time, I have set the setting in my personal games to 255 and not touched it since. I find the spindly winding rivers quite aesthetically pleasing. Default is the original hardcoded value of 5.

Control the size of lakes: This can look uglier on larger settings since the sides of lakes closest to the 'feed point' can be very straight, but I have also seen this behavior exhibited in real life. Setting is 1-255, I find 50 is a nice comfortable setting for the games I personally play.

Control whether lakes can generate in deserts: This seemed like a silly restriction, as a tropic zone gets generated around the lake regardless I felt there was no harm in allowing their generation.

### CODE REVIEW

File: src/table/settings.ini [hunk 1]
```diff
@@ -3721,21 +3721,75 @@ cat      = SC_BASIC
...
+patxname = ""rainforest_line_height.game_creation.rainforest_line_height""
+
+[SDT_VAR]
+base     = GameSettings
+var      = game_creation.lake_size
+type     = SLE_UINT8
+guiflags = SGF_NEWGAME_ONLY | SGF_SCENEDIT_TOO
+def      = 3
+min      = 1
+max      = 255
```
Comment (JGRennison) [inline src/table/settings.ini:3779]: Would it be worth constraining the upper limit a bit?
Diameter 255 lakes seems potentially overkill.

File: src/landscape.cpp [hunk 2]
```diff
@@ -1266,15 +1269,15 @@ static bool FlowRiver(TileIndex spring, TileIndex begin)
...
 				/* We only want a lake if the river is long enough. */
 				DistanceManhattan(spring, lakeCenter) > _settings_game.game_creation.min_river_length) {
 			end = lakeCenter;
 			MakeRiver(lakeCenter, Random());
 			/* Remove desert directly around the river tile. */
-			CircularTileSearch(&lakeCenter, RIVER_OFFSET_DESERT_DISTANCE, RiverModifyDesertZone, nullptr);
+			CircularTileSearch(&lakeCenter, _settings_game.game_creation.river_tropics_width, RiverModifyDesertZone, nullptr);
 			lakeCenter = end;
-			uint range = RandomRange(8) + 3;
+			uint range = RandomRange(8) + _settings_game.game_creation.lake_size;
```
Comment (JGRennison) [inline src/landscape.cpp:1280]: Similar to the other PR, I wonder if the RandomRange argument should also be a function of the lake size setting.

File: src/table/settings.ini (hunk 1 above)
Comment (reldred) [reply src/table/settings.ini:3779]: I'm not against it. I play with 50 personally, does 100 seem like a reasonable compromise?

File: src/landscape.cpp (hunk 2 above)
Comment (reldred) [reply src/landscape.cpp:1280]: I'll experiment and see how this affects generation.

File: src/table/settings.ini (hunk 1 above)
Comment (reldred) [reply src/table/settings.ini:3779]:
```suggestion
max      = 100
```

Comment (JGRennison) [reply src/table/settings.ini:3779]: That should be fine

File: src/landscape.cpp (hunk 2 above)
Comment (reldred) [reply src/landscape.cpp]: > 
> 
> Similar to the other PR, I wonder if the RandomRange argument should also be a function of the lake size setting.

While it didn't work well for the other PR, it worked out fine here. I've committed a new change to make the max size of the random range our new setting, changed the default value of the setting to 8 to match the original hard coded value. Now small lakes can generate as well as large lakes. Still not a fan of the shape of the lakes in larger sizes but that's a challenge for another day.

**Comment by JGRennison** [comment]
Many thanks for this. I've squashed and applied this with some minor adjustments, it'll be in the next release.

**Pull request closed without merging**